server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java

```diff
@@ -260,13 +268,19 @@ public DesiredBalance compute(
                 break;
             }
             if (i % 100 == 0) {
-                // TODO this warning should be time based, iteration count should be proportional to the number of shards
+                final int iterations = i;
+                final long currentTime = System.nanoTime();
+                final boolean shouldWarnByTime = nextWarningTime <= currentTime;
+                if (shouldWarnByTime) {
+                    nextWarningTime += timeWarningInterval;
```

Comment:
I think we should advance the time according to `currentTime`, otherwise (e.g. if the process pauses) we might get many warnings in quick succession.

```suggestion
                    nextWarningTime = currentTime + timeWarningInterval;
```